base/libgit2/types.jl

```diff
@@ -754,50 +735,31 @@ function securezero!(cred::SSHCredentials)
     securezero!(cred.pass)
     securezero!(cred.pubkey)
     securezero!(cred.prvkey)
+    cred.count = 0
     return cred
 end
 
 "Credentials that support caching"
 type CachedCredentials <: AbstractCredentials
-    cred::Dict{String,SSHCredentials}
+    cred::Dict{String,AbstractCredentials}
     count::Int            # authentication failure protection count
-    CachedCredentials() = new(Dict{String,SSHCredentials}(),3)
-end
-"Returns specific credential parameter value: first index is a credential
-parameter name, second index is a host name (with schema)"
-function Base.getindex(p::CachedCredentials, keys...)
-    length(keys) != 2 && return nothing
-    key, host = keys
-    if haskey(p.cred, host)
-        creds = p.cred[host]
-        if isdefined(creds,key)
-            kval = getfield(creds, key)
-            !isempty(kval) && return kval
-        end
-    end
-    return nothing
-end
-"Sets specific credential parameter value: first index is a credential
-parameter name, second index is a host name (with schema)"
-function Base.setindex!(p::CachedCredentials, val, keys...)
-    length(keys) != 2 && return nothing
-    key, host = keys
-    if !haskey(p.cred, host)
-        p.cred[host] = SSHCredentials()
-    end
-    creds = p.cred[host]
-    @assert isdefined(creds,key)
-    setfield!(creds, key, val)
-    return p
+    CachedCredentials() = new(Dict{String,AbstractCredentials}(),3)
 end
+
 "Checks if credentials were used or failed authentication, see `LibGit2.credentials_callback`"
-function checkused!(p::CachedCredentials)
+function checkused!(p::Union{UserPasswordCredentials, SSHCredentials})
     p.count <= 0 && return true
     p.count -= 1
     return false
 end
-"Resets authentication failure protection count"
-reset!(p::CachedCredentials, cnt::Int=3) = (p.count = cnt)
+reset!(p::Union{UserPasswordCredentials, SSHCredentials}, cnt::Int=3) = (p.count = cnt)
+reset!(p::CachedCredentials) = foreach(reset!, values(p.cred))
```

Comment:
I'm not really sure what the point of retrying 3 times is anyway. The protocol is generally realiable and transport errors won't call the authentication callback anyway. I kept it this way because that's what was there before, but this is probably something to get rid of anyway.